### PR TITLE
Emitter: guard _RAND_* declarations with ifdef

### PR DESCRIPTION
**Pull request by albertchen-sifive**
fixes https://github.com/freechipsproject/firrtl/issues/1372. This adds `ifdefInitials` and `ifdefDeclares` fields to `VerilogRenderer` which map ifdef strings to their declaration/initial blocks and adds versions of `declare`, `declareVectorType`, and `rand_string` that take an optional ifdef string.

### Contributor Checklist

- [x] Did you add Scaladoc to every public function/method?
- [ ] Did you update the FIRRTL spec to include every new feature/behavior?
- [ ] Did you add at least one test demonstrating the PR?
- [x] Did you delete any extraneous printlns/debugging code?
- [x] Did you specify the type of improvement?
- [x] Did you state the API impact?
- [x] Did you specify the code generation impact?
- [x] Did you request a desired merge strategy?
- [x] Did you add text to be included in the Release Notes for this change?

#### Type of Improvement

- backend code generation

#### API Impact

extends the declare/declareVectorType/rand_string methods in `VerilogRender` to include versions that accept an ifdef string.

#### Backend Code Generation Impact

`_RAND_*` declarations are gated by `ifdef`s and grouped together. Instead of having `_RAND_*` regs mixed with normal regs the `_RAND_*` regs will be grouped into ifdef blocks. Initialization code for `_RAND_*` regs will also be grouped by ifdefs. This is what ICache.v looks like
```
module ICache(
...
);
`ifdef RANDOMIZE_MEM_INIT
  reg [31:0] _RAND_0;
  reg [31:0] _RAND_3;
...
`endif // RANDOMIZE_MEM_INIT
`ifdef RANDOMIZE_REG_INIT
  reg [31:0] _RAND_1;
  reg [31:0] _RAND_2;
...
`endif // RANDOMIZE_REG_INIT
  reg [19:0] tag_array_0 [0:63];
  wire [19:0] tag_array_0_tag_rdata_data;
...
initial begin
  `ifdef RANDOMIZE
    `ifdef INIT_RANDOM
      `INIT_RANDOM
    `endif
    `ifndef VERILATOR
      `ifdef RANDOMIZE_DELAY
        #`RANDOMIZE_DELAY begin end
      `else
        #0.002 begin end
      `endif
    `endif
`ifdef RANDOMIZE_MEM_INIT
  _RAND_0 = {1{`RANDOM}};
  for (initvar = 0; initvar < 64; initvar = initvar+1)
    tag_array_0[initvar] = _RAND_0[19:0];
  _RAND_3 = {1{`RANDOM}};
  for (initvar = 0; initvar < 64; initvar = initvar+1)
    tag_array_1[initvar] = _RAND_3[19:0];
  ...
`endif // RANDOMIZE_MEM_INIT
`ifdef RANDOMIZE_REG_INIT
  _RAND_1 = {1{`RANDOM}};
  tag_array_0_tag_rdata_en_pipe_0 = _RAND_1[0:0];
  _RAND_2 = {1{`RANDOM}};
  tag_array_0_tag_rdata_addr_pipe_0 = _RAND_2[5:0];
  ...
`endif // RANDOMIZE_REG_INIT
  `endif // RANDOMIZE
end // initial
...
```

and this is what the old ICache looked like
```
module ICache(
...
)
  reg [19:0] tag_array_0 [0:63];
  reg [31:0] _RAND_0;
  wire [19:0] tag_array_0_tag_rdata_data;
  wire [5:0] tag_array_0_tag_rdata_addr;
  wire [19:0] tag_array_0__T_328_data;
  wire [5:0] tag_array_0__T_328_addr;
  wire  tag_array_0__T_328_mask;
  wire  tag_array_0__T_328_en;
  reg  tag_array_0_tag_rdata_en_pipe_0;
  reg [31:0] _RAND_1;
  reg [5:0] tag_array_0_tag_rdata_addr_pipe_0;
  reg [31:0] _RAND_2;
  reg [19:0] tag_array_1 [0:63];
  reg [31:0] _RAND_3;
...
initial begin
  `ifdef RANDOMIZE
    `ifdef INIT_RANDOM
      `INIT_RANDOM
    `endif
    `ifndef VERILATOR
      `ifdef RANDOMIZE_DELAY
        #`RANDOMIZE_DELAY begin end
      `else
        #0.002 begin end
      `endif
    `endif
  _RAND_0 = {1{`RANDOM}};
  `ifdef RANDOMIZE_MEM_INIT
  for (initvar = 0; initvar < 64; initvar = initvar+1)
    tag_array_0[initvar] = _RAND_0[19:0];
  `endif // RANDOMIZE_MEM_INIT
  `ifdef RANDOMIZE_REG_INIT
  _RAND_1 = {1{`RANDOM}};
  tag_array_0_tag_rdata_en_pipe_0 = _RAND_1[0:0];
  `endif // RANDOMIZE_REG_INIT
  `ifdef RANDOMIZE_REG_INIT
  _RAND_2 = {1{`RANDOM}};
  tag_array_0_tag_rdata_addr_pipe_0 = _RAND_2[5:0];
  `endif // RANDOMIZE_REG_INIT
  _RAND_3 = {1{`RANDOM}};
  `ifdef RANDOMIZE_MEM_INIT
...
  `endif // RANDOMIZE
end // initial
...
```
I wasn't sure about what the indentation should look like. right now `ifdef`/`endif` lines are not indented and the statements inside are indented once.

#### Desired Merge Strategy

- Squash: The PR will be squashed and merged (choose this if you have no preference.

#### Release Notes

Guard `_RAND_*` randomization registers with `ifdef`s.

### Reviewer Checklist (only modified by reviewer)
- [ ] Did you add the appropriate labels?
- [ ] Did you mark the proper milestone (1.2.x, 1.3.0, 1.4.0) ?
- [ ] Did you review?
- [ ] Did you check whether all relevant Contributor checkboxes have been checked?
- [ ] Did you mark as `Please Merge`?

### CODE REVIEW

File: src/main/scala/firrtl/Emitter.scala [hunk 1]
```diff
@@ -937,7 +972,12 @@ class VerilogEmitter extends SeqTransform with Emitter {
       for (x <- portdefs) emit(Seq(tab, x))
       emit(Seq(");"))
 
-      if (declares.isEmpty && assigns.isEmpty) emit(Seq(tab, "initial begin end"))
+      ifdefDeclares.toSeq.sortWith(_._1 < _._1).foreach { case (ifdef, declares) =>
+        emit(Seq("`ifdef " + ifdef))
+        for (x <- declares) emit(Seq(tab, x))
+        emit(Seq("`endif // " + ifdef))
+      }
+      if (declares.isEmpty && ifdefDeclares.isEmpty && assigns.isEmpty) emit(Seq(tab, "initial begin end"))
```
[the remark this answers is not in the document]
Comment (mwachs5) [reply src/main/scala/firrtl/Emitter.scala:980]: i think you do NOT want to, I've heard complaints about empty initial blocks

File: src/main/scala/firrtl/Emitter.scala [hunk 2]
```diff
@@ -544,6 +544,8 @@ class VerilogEmitter extends SeqTransform with Emitter {
     }
 
     val portdefs = ArrayBuffer[Seq[Any]]()
+    // mapps ifdef guard to declaration blocks
+    val ifdefDeclares = mutable.Map[String, ArrayBuffer[Seq[Any]]]()
```
[the remark this answers is not in the document]
Comment (albertchen-sifive) [reply src/main/scala/firrtl/Emitter.scala:548]: thanks! did not know that was a thing

File: src/main/scala/firrtl/Emitter.scala [hunk 3]
```diff
@@ -576,31 +580,56 @@ class VerilogEmitter extends SeqTransform with Emitter {
...
+      }
     }
-    def declareVectorType(b: String, n: String, tpe: Type, size: BigInt, info: Info, preset: Expression) = {
+
+    // original vector type declare without initial value
+    def declareVectorType(b: String, n: String, tpe: Type, size: BigInt, info: Info): Unit =
+      declareVectorType(b, n, tpe, size, info, None)
+
+    // declare vector type with initial value
+    def declareVectorType(b: String, n: String, tpe: Type, size: BigInt, info: Info, preset: Expression): Unit = {
```
[the remark this answers is not in the document]
Comment (albertchen-sifive) [reply src/main/scala/firrtl/Emitter.scala:598]: I can give them different names. I thought these were all public though?

[the remark this answers is not in the document]
Comment (seldridge) [reply src/main/scala/firrtl/Emitter.scala:598]: My mistake. I backwards searched to a different scope. I think I would just make the new methods private.  Another PR (out of scope for this PR) can start deprecating things in `VerilogRender`. This thing has a massive API surface for no real reason.

File: src/main/scala/firrtl/Emitter.scala (hunk 1 above)
[the remark this answers is not in the document]
Comment (albertchen-sifive) [reply src/main/scala/firrtl/Emitter.scala]: I removed it

File: src/main/scala/firrtl/Emitter.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (albertchen-sifive) [reply src/main/scala/firrtl/Emitter.scala:606]: ok. they are private now

**Comment by jackkoenig** [comment]
> I'll just wait for the binary compatibility to be setup for this before merging. Thanks!

And wait for the automated backporting 🙂

**Comment by seldridge** [comment]
Ack. Automated backporting is what I meant...

**Comment by jackkoenig** [comment]
Done!

**Pull request merged**